Add support for loading files, decoding them as JSONL, and formatting them using a Logback-like format string.

junhaoliao opened the pull request:
# References
#45

# Description

1. Implement LogFileManager class for efficient log file handling.
2. Add decoder interface.
3. Add Decode Options to File Loading Process.
4. Add support for JSONL log file decoding.

# Validation performed

1. Placed example.jsonl file under `/public/test`. (Can be downloaded and extracted from [example.jsonl.zip](https://github.com/user-attachments/files/15844173/example.jsonl.zip))
2. Launched debug server with `npm run start`.
3. Observed last page of decoded results displayed in the browser. e.g.:
   ![image](https://github.com/y-scope/yscope-log-viewer/assets/43196707/264ac0d1-b384-4825-9081-a6a21e6d30c8)

junhaoliao committed:
Add log file management and jsonl decoder functionality.

junhaoliao commented:
@Henry8192 Please help triage the review.

junhaoliao committed:
Add docs to Decoder interface.

junhaoliao committed:
Add decode options to file loading process.

junhaoliao committed:
Revert changes unrelated to current PR.

junhaoliao committed:
Add JSON value and object types annotations.

Henry8192 commented on new-log-viewer/src/services/decoders/JsonlDecoder.ts:
```diff
@@ -0,0 +1,240 @@
+        if (null === dateFormatMatch) {
+            console.warn(
+                "Unable to find date format string in #textPattern:",
+                this.#textPattern
```
```suggestion
    /**
    * Sets the text pattern for the decoder and extracts the date format and property names from it.
    *
    * @param {string} pattern - The text pattern to set for the decoder.
    * @return {void}
    */
    #setTextPattern (pattern: string) {
```

junhaoliao replied:
For functions returning `void` (doesn't return a value), we can omit the `@return` in the docs.

Henry8192 commented on new-log-viewer/src/services/decoders/JsonlDecoder.ts:
```diff
@@ -0,0 +1,240 @@
+        } else {
+            verbosity = LOG_VERBOSITY[verbosityString as (keyof typeof LOG_VERBOSITY)];
+        }
+
```
```suggestion
    /**
    * Builds an index by decoding the data array and splitting it into lines.
    * Each line is parsed as a JSON object and added to the log events array.
    * If a line cannot be parsed as a JSON object, an error is logged and the line is skipped.
    *
    * @return {number} The number of log events in the log events array after building the index.
    */
    buildIdx (): number {
```

Henry8192 commented on new-log-viewer/src/services/decoders/JsonlDecoder.ts:
```diff
@@ -0,0 +1,240 @@
+
+    buildIdx (): number {
+        const text = JsonlDecoder.TEXT_DECODER.decode(this.#dataArray);
+        const split = text.split("\n");
```
```suggestion
    /**
    * Decodes log events from the #logEvents array and adds them to the results array.
    *
    * @param {DecodeResultType[]} results - The array to which the decoded log events will be added.
    * @param {number} startIdx - The index in the #logEvents array at which to start decoding.
    * @param {number} endIdx - The index in the #logEvents array at which to stop decoding.
    * @return {boolean} - Returns true if the decoding was successful, false otherwise.
    */
    decode (results: DecodeResultType[], startIdx: number, endIdx: number): boolean {
```

Henry8192 commented on new-log-viewer/src/services/LogFileManager.ts:
```diff
@@ -0,0 +1,138 @@
+        this.#pageSize = pageSize;
+    }
+
+    async loadFile (fileSrc: FileSrcType): Promise<number> {
```
```suggestion
    /**
    * Loads a file from a given source and decodes it using the appropriate decoder based on the file extension.
    *
    * @param {FileSrcType} fileSrc - The source of the file to load. This can be a string representing a URL, or a File object.
    * @return {Promise<number>} - A promise that resolves with the number of log events found in the file.
    * @throws {Error} - Throws an error if the file source type is not supported.
    */
    async loadFile (fileSrc: FileSrcType): Promise<number> {
```

junhaoliao replied:
> @throws {Error} - Throws an error if the file source type is not supported.

I guess we need to change Line 64 to throw instead, right?

Henry8192 commented on new-log-viewer/src/services/LogFileManager.ts:
```diff
@@ -0,0 +1,138 @@
+        this.#decoder.setDecodeOptions(options);
+    }
+
+    loadPage (cursor: CursorType): {
```
```suggestion
    /**
    * Loads a page of log events based on the provided cursor.
    *
    * @param {CursorType} cursor - The cursor indicating the page to load. This can be a page number or other types of cursor.
    * @returns {Object} An object containing the logs as a string, a map of line numbers to log event numbers, and the cursor line number.
    * @throws {Error} - Throws an error if the loadFile method has not been called before this method.
    */
    loadPage (cursor: CursorType): {
```

junhaoliao replied:
How about

/**
 * Loads a page of log events based on the provided cursor.
 *
 * @param cursor The cursor indicating the page to load. See {@link CursorType}.
 * @returns An object containing the logs as a string, a map of line numbers to log event numbers, and the line number of the first line in the cursor identified event.
 * @throws {Error} - Throws an error if the loadFile method has not been called before this method.
 */

Henry8192 reviewed (commented):
Added some JSDocs.

junhaoliao commented on new-log-viewer/src/components/Layout.tsx:
```diff
@@ -15,16 +15,27 @@ const Layout = () => {
 
     useEffect(() => {
-        loadFile("");
+        loadFile("http://localhost:3010/test/example.jsonl");
```
@kirkrodrigues What do you think? Shall we include example files for testing purpose?

junhaoliao commented on new-log-viewer/src/services/decoders/JsonlDecoder.ts:
```diff
@@ -0,0 +1,230 @@
+        return dateFormat;
+    }
+
+    #extractDateFormat () {
```
For private functions (either explicitly marked as `private` in TypeScript or we prefix their names with `#`), we do not need to mark them as `@private` in the docs.

junhaoliao commented on new-log-viewer/src/utils/http.ts:
```diff
@@ -0,0 +1,71 @@
+    );
+};
+
+
```
```suggestion
```

kirkrodrigues commented on new-log-viewer/src/services/LogFileManager.ts:
```diff
@@ -0,0 +1,138 @@
+ * A mapping of file name extensions to their respective decoders.
+ */
+const FILE_EXT_TO_DECODER: Readonly<Record<string, DecoderConstructor>> = Object.freeze({
+    ".jsonl": JsonlDecoder as unknown as DecoderConstructor,
```
Rather than breaking out of the type system, could we just handle all the cases in the `loadFile`? Is there anywhere else where we would need to determine the decoderconstructor at runtime?

junhaoliao replied:
Not sure if I understood the question correctly. The goal of storing a map of file extensions to decoder constructors is to allow different types of decoders to be created depending on the file name extension. Note:
1. The method signature of some class's constructor must be separate from the class's definition, as done in Decoder.ts.
2. Eventually we will provide a magic number based file type detection mechanism.

junhaoliao replied:
Discussed offline and we concluded that inlining the cases into the code would be better.

kirkrodrigues commented on new-log-viewer/src/services/LogFileManager.ts:
```diff
@@ -0,0 +1,138 @@
+        this.#pageSize = pageSize;
+    }
+
+    async loadFile (fileSrc: FileSrcType): Promise<number> {
```
Can we differentiate an empty file from a failure to load it?

kirkrodrigues commented on new-log-viewer/src/services/LogFileManager.ts:
```diff
@@ -0,0 +1,138 @@
+        const results: Array<[string, number, number, number]> = [];
+        let startLogEventNum = (Math.floor(this.#numEvents / this.#pageSize) * this.#pageSize) + 1;
+        if (null !== cursor) {
+            if ("pageNum" in cursor) {
```
I haven't thought this through, but rather than type-checking like this, could we do something like what you did for the request/response? I.e., each cursor type has a type code and the arguments for the cursor type depend on the type code.

junhaoliao replied:
Right, the type code is required for the generic function to work. Let me look into that.

kirkrodrigues commented on new-log-viewer/src/services/LogFileManager.ts:
```diff
@@ -0,0 +1,138 @@
+        let currentLine = 1;
+        results.forEach((r) => {
+            const [
+                m,
```
```suggestion
                message,
```

Not absolutely necessary, but `message` or `msg` seems like a free improvement.

kirkrodrigues commented on new-log-viewer/src/services/LogFileManager.ts:
```diff
@@ -0,0 +1,138 @@
+
+            messages.push(m);
+            lines.set(currentLine, logEventNum);
+            currentLine += m.split("\n").length - 1;
```
Something to think about for the future---perhaps we should return the number of lines per log event from the decoder.

junhaoliao replied:
Hmm.. how do we plan to use that? With that approach, we would have to use an accumulator to deduct the log event Num (index) from the editor cursor line, right? (Do you mean we send the lines per log event to the renderer, and let the renderer calculate once and cache the `beginLineNumToLogEventNum[]`?)

kirkrodrigues replied:
I just mean we could use it here instead of calling `msg.split('\n').length - 1` every time we load a page.

junhaoliao commented on new-log-viewer/src/contexts/StateContextProvider.tsx:
```diff
@@ -82,7 +82,7 @@ const StateContextProvider = ({children}: StateContextProviderProps) => {
                 setLogData(args.logs);
                 setLogLines(args.lines);
-                setLogEventNum(args.startLogEventNum);
+                setLogEventNum(args.lines.get(1) as number);
```
* Sorry for the confusions. `logEventNum` was intended to replace `logEventIdx` to be consistent with the new naming convention, since the index starts from 1 rather than 0.
* Sure, I agree `beginLineNumToLogEventNum` is a better name.

kirkrodrigues replied:
Hm, so if `logEventNum` is meant to be the log event requested by the user, is `setLogEventNum(args.lines.get(1) as number);` just meant to set the log event number to the first log event in the page? Why would we want to do that?

junhaoliao replied:
It was a placeholder before Monaco can be integrated. To show the last event when the log file gets loaded, it should be this instead:
```
const lineNumbers = Array.from(args.beginLineNumToLogEventNum.values());
const lastLogEventNum = lineNumbers[args.beginLineNumToLogEventNum.size - 1];
if ("undefined" !== typeof lastLogEventNum) {
    setLogEventNum(lineNumbers[args.beginLineNumToLogEventNum.size - 1] as number);
} else {
    console.error("Last logEventNum expectedly undefined");
}
```

kirkrodrigues replied:
Gotcha. But if the user requests a certain logEventNum, we should seek and show that one specifically rather than the last log event on the page, right?

junhaoliao commented on new-log-viewer/src/contexts/StateContextProvider.tsx:
```diff
@@ -32,7 +32,7 @@ const StateContext = createContext<StateContextType>({} as StateContextType);
 const StateDefaultValue = Object.freeze({
     loadFile: () => null,
-    logData: "",
+    logData: "Loading...",
```
Sure. A "Notification Context Provider", which allow children components to post notifications (/ debug logs), will be submitted with a future "Monaco Integration" PR.

junhaoliao committed:
Allow user to specify file path in URL.

junhaoliao committed:
Apply suggestions from code review

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

junhaoliao commented on new-log-viewer/src/components/Layout.tsx:
```diff
@@ -15,16 +18,33 @@ const Layout = () => {
             <div>
-                {logData}
+                <h3>
+                    LogEventNum -
```
Sorry for the confusion. Let's see the discussion at https://github.com/y-scope/yscope-log-viewer/pull/46/files#r1643014357

junhaoliao committed:
Rename 'logLines' to 'beginLineNumToLogEventNum'.

junhaoliao committed:
Refactor field and variables related to log 'verbosity' to 'level' & 'propName' to 'key'.

junhaoliao committed:
Update last log event number logic in StateContextProvider.

junhaoliao committed:
Apply suggestions from code review

Co-authored-by: Henry8192 <50559854+Henry8192@users.noreply.github.com>

junhaoliao committed:
Add error handling.

junhaoliao committed:
Refactor decoder object in src/services and src/typings.

junhaoliao committed:
Apply suggestions from code review

Co-authored-by: Henry8192 <50559854+Henry8192@users.noreply.github.com>

junhaoliao committed:
Refactor loadPage docs and variable names in LogFileManager.

junhaoliao committed:
Typify cursor handling.

junhaoliao committed:
Rename global constant name from 'StateDefaultValue' to 'STATE_DEFAULT'.

junhaoliao committed:
Apply suggestions from code review

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

junhaoliao committed:
Apply suggestions from code review

Co-authored-by: Henry8192 <50559854+Henry8192@users.noreply.github.com>

junhaoliao committed:
Apply suggestions from code review

Co-authored-by: Henry8192 <50559854+Henry8192@users.noreply.github.com>

junhaoliao committed:
Apply suggestions from code review

Co-authored-by: Henry8192 <50559854+Henry8192@users.noreply.github.com>

junhaoliao committed:
Refactor JsonlDecoder and improve comments.

junhaoliao commented on new-log-viewer/src/services/decoders/JsonlDecoder.ts:
```diff
@@ -0,0 +1,240 @@
+        return true;
+    }
+
+    buildIdx (): number {
```
Not necessarily. If we enable the optimization in the future where we allow building partial index wnen the decoder `.decode([], startIdx=1, endIdx=any)`. `buildIdx()` can be called after the first page is decoded.

junhaoliao replied:
Yes.

With some other thoughts, I also agree we want to hide this interface. For now, We can make `buildIdx()` a helper of `decode()` to be invoked conditionally. Then when we work on the optimization we can extract the core logics from `buildIdx()` to allow those to be reused.

junhaoliao replied:
Sorry, we might not want to do that since we do need the total number of events to see where's the last page, before we can introduce the optimization where we skip building index for the whole log file when only the first page is requested.

junhaoliao replied:
> for CLP formats, I would deserialize + decode the preamble in the constructor.

Sure. Let me correct the docs.

junhaoliao committed:
Apply suggestions from code review

Co-authored-by: Henry8192 <50559854+Henry8192@users.noreply.github.com>

junhaoliao commented on new-log-viewer/src/services/decoders/JsonlDecoder.ts:
```diff
@@ -0,0 +1,240 @@
+        for (let logEventIdx = startIdx; logEventIdx < endIdx; logEventIdx++) {
+            const logEvent = this.#logEvents[logEventIdx];
+            if ("undefined" === typeof logEvent) {
+                return false;
```
> How could this happen? Are you anticipating an OOB error?

I don't anticipate this to happen although the subscripting operator does return "undefined" when such happens. 

> If so, couldn't we check the range is within the bounds of #logEvents?

Unfortunately, TypeScript hasn't been smart enough to assume the non-`undefined` type even with explicit range checking. 

(With TypeScript not being smart, imagine how much overhead it would introduce to implement an algorithm like Binary Search, lol.)

kirkrodrigues replied:
Hm, in that case, would using a slice avoid typescript's complaints?

junhaoliao commented on new-log-viewer/src/services/decoders/JsonlDecoder.ts:
```diff
@@ -0,0 +1,240 @@
+        return this.#logEvents.length;
+    }
+
+    decode (results: DecodeResultType[], startIdx: number, endIdx: number): boolean {
```
The interface was designed with considerations for our Emscripten compiled decoders. Creating an output array in Emscripten isn't that straightforward but seems possible and with low overhead. I'll make the change now, and if not doable in Emscripten we can take a look again.

junhaoliao replied:
Correction - actually fairly straight forward as now I find this in the doc: https://emscripten.org/docs/api_reference/val.h.html#_CPPv4N10emscripten10emscripten3val5arrayEv

junhaoliao commented on new-log-viewer/src/services/decoders/JsonlDecoder.ts:
```diff
@@ -0,0 +1,240 @@
+                this.#logEvents.push(logEvent);
+            } catch (e) {
+                if (e instanceof SyntaxError) {
+                    console.error(e, line);
```
I missed the `else{throw ...}` but [according to the docs no exceptions should be thrown](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#exceptions).

junhaoliao committed:
Apply suggestions from code review

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

junhaoliao commented on new-log-viewer/src/utils/url.ts:
```diff
@@ -0,0 +1,23 @@
+        const url = new URL(urlString);
+        const parts = url.pathname.split("/");
+        const popped = parts.pop();
+        if ("string" === typeof popped) {
```
"undefined"

kirkrodrigues replied:
How come?

junhaoliao replied:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/pop#return_value

pop() returns `undefined` when the array is empty. TypeScript is not smart enough to tell if an array is empty or not.

Since we're not logging `popped` being unexpectedly `undefined` when such happens, alternatively we can cast the return type in Line 16:
```
const popped = parts.pop() as string;
```

kirkrodrigues replied:
Gotcha. Perhaps we should add a comment when we're using unnecessary code to satisfy the typescript compiler. On that note, I do prefer the cast you suggested.

junhaoliao commented on new-log-viewer/src/utils/url.ts:
```diff
@@ -0,0 +1,23 @@
+ * Extracts the basename (filename) from a given URL string.
+ *
+ * @param urlString
+ * @return The extracted basename or "Unknown Filename" if extraction fails.
```
> if you want to avoid making the caller handle the error

Right

> why not pass in a default value?

Sounds better

junhaoliao commented on new-log-viewer/src/services/decoders/JsonlDecoder.ts:
```diff
@@ -0,0 +1,240 @@
+
+    buildIdx (): number {
+        const text = JsonlDecoder.TEXT_DECODER.decode(this.#dataArray);
+        const split = text.split("\n");
```
Mind elaborating? The `split()` JS API has long been known for its efficiency due to V8 optimizations, but I guess not in all cases especially when we have a large string.

junhaoliao replied:
hmm... Mind showing your benchmarking code? With below `buildIdx ()` constantly takes 1400-1500ms while the original code takes 1300-1400ms.

```
    buildIdx (): number {
        const text = `${JsonlDecoder.#textDecoder.decode(this.#dataArray)}\n`;
        let startIndex = 0;
        let endIndex = text.indexOf("\n");

        while (-1 !== endIndex) {
            const line = text.slice(startIndex, endIndex);
            if (0 < line.length) {
                try {
                    const logEvent = JSON.parse(line) as JsonObject;
                    this.#logEvents.push(logEvent);
                } catch (e) {
                    console.error(e, line);
                }
            }
            startIndex = endIndex + 1;
            endIndex = text.indexOf("\n", startIndex);
        }

        return this.#logEvents.length;
    }
```

kirkrodrigues replied:
```js
    buildIdx (): number {
        const text = JsonlDecoder.#textDecoder.decode(this.#dataArray);

        const beginTimestamp = performance.now();

        const split = text.split("\n");
        for (const line of split) {
            if (0 === line.length) {
                continue;
            }

            try {
                const logEvent = JSON.parse(line) as JsonObject;
                this.#logEvents.push(logEvent);
            } catch (e) {
                console.error(e, line);
            }
        }

        // let beginIdx = 0;
        // while (beginIdx < text.length) {
        //     const endIdx = text.indexOf("\n", beginIdx);
        //     let line;
        //     if (-1 === endIdx) {
        //         line = text.substring(beginIdx);
        //         beginIdx = text.length;
        //     } else {
        //         line = text.substring(beginIdx, endIdx);
        //         beginIdx = endIdx + 1;
        //     }
        //     if (0 === line.length) {
        //         continue;
        //     }
        //
        //     try {
        //         const logEvent = JSON.parse(line) as JsonObject;
        //         this.#logEvents.push(logEvent);
        //     } catch (e) {
        //         console.error(e, line);
        //     }
        // }

        const endTimestamp = performance.now();
        console.log(endTimestamp - beginTimestamp, "ms");

        return this.#logEvents.length;
    }
```

Compared the two implementations above on a 94MB log file.

junhaoliao replied:
Your `indexOf` implementation is indeed slightly faster than mine. When I was testing with the 512MB log file, the speed up compared the my original `split` implementation was about 30ms with the `performance.now()` counters. 

However, after I removed the `performance.now()` counters and used the browser profiler (Debugger - Performance tab) to measure the time, both your `indexOf` implementation and my `split` seem to take 1.32 ~ 1.33 s consistently. I did notice the JS heap size is 8MB less (468MB v.s. 476MB) with your `indexOf`, which makes sense as we previously suspect `split` could be caching chunks of split results to speed up the accesses just in time.

I imagine the results could be vastly different if we have some very long lines in the test files.

With that said, both implementations are concise and should be easy to maintain. Which you do you prefer?

junhaoliao replied:
In both implementations, I propose we can move the `if (0 === line.length)` check in the `catch` clause since it's uncommon to have empty lines in the middle of such log files. The performance speed up with the 512M file was about 5ms. e.g.,
```
            try {
                const logEvent = JSON.parse(line) as JsonObject;
                this.#logEvents.push(logEvent);
            } catch (e) {
                if (0 === line.length) {
                    console.error(e, line);
                }
            }
```

junhaoliao committed:
Apply suggestions from code review

Co-authored-by: Henry8192 <50559854+Henry8192@users.noreply.github.com>

junhaoliao committed:
Refactor decoding functions and error handling.

junhaoliao commented on new-log-viewer/src/typings/decoders.ts:
```diff
@@ -0,0 +1,15 @@
+/**
+ * Type of decode options passed to the decoder before calling their decode() method.
+ */
+type DecodeOptionsType = JsonlDecodeOptionsType;
```
I think the options for all decoders can directly come from URL parameters and/or localStorage. When any of those are missing for a particular decoder, we provide default values. 

It make sense from a config standpoint that not all configs are useful. And the type definition here is only for information purpose.

junhaoliao commented on new-log-viewer/src/utils/url.ts:
```diff
@@ -0,0 +1,23 @@
+    let basename = "Unknown Filename";
+    try {
+        const url = new URL(urlString);
+        const parts = url.pathname.split("/");
```
Right, we expect the urlString here is url decoded. Let me put that into the docs.

junhaoliao committed:
Update URL parsing functionality and improve error reporting.

junhaoliao committed:
Apply suggestions from code review

junhaoliao committed:
Refactor "textPattern" to "formatString" in JsonlDecoder.

junhaoliao committed:
Merge remote-tracking branch 'junhao/new-log-viewer' into new-log-viewer

junhaoliao committed:
Fix typo.

junhaoliao committed:
Reorganize decoder and log file manager classes.

kirkrodrigues commented on new-log-viewer/src/utils/url.ts:
```diff
@@ -0,0 +1,27 @@
+            basename = popped;
+        }
+    } catch (e) {
+        console.error("Error happened in parsing file name", e);
```
```suggestion
        console.error("Failed to parse basename.", e);
```

kirkrodrigues commented on new-log-viewer/src/utils/http.ts:
```diff
@@ -0,0 +1,70 @@
+};
+
+/**
+ * Downloads and reads a file as a Uint8Array with a given URL.
```
```suggestion
 * Downloads (bypassing any caching) and reads a file as a Uint8Array with a given URL.
```

kirkrodrigues commented on new-log-viewer/src/utils/http.ts:
```diff
@@ -0,0 +1,70 @@
+ * @param fileUrl
+ * @param progressCallback Callback to update progress
+ * @return File content
+ * @throws {Error}
```
```suggestion
 * @throws {Error} if the download fails.
```

kirkrodrigues commented on new-log-viewer/src/utils/url.ts:
```diff
@@ -0,0 +1,27 @@
+ *
+ * @param urlString a URL string that does not contain escaped `/` (%2F).
+ * @param defaultFileName
+ * @return The extracted basename or "Unknown Filename" if extraction fails.
```
```suggestion
 * @return The extracted basename or `defaultFileName` if extraction fails.
```

kirkrodrigues reviewed (changes requested):
Some more suggestions and design questions. There are also some comments from the previous review that I think might've been missed.

junhaoliao committed:
Apply suggestions from code review

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

junhaoliao commented on new-log-viewer/src/services/decoders/JsonlDecoder.ts:
```diff
@@ -0,0 +1,269 @@
+ */
+type JsonObject = { [key: string]: JsonValue };
+
+class JsonlDecoder implements Decoders {
```
Sure. Let me work on that quickly.

junhaoliao commented on new-log-viewer/src/services/decoders/JsonlDecoder.ts:
```diff
@@ -0,0 +1,269 @@
+     * @param dateFormat The Logback date format string to convert.
+     * @return The corresponding Day.js date format string.
+     */
+    static #convertLogbackDateFormatToDayjs (dateFormat: string) {
```
This was assuming the user would supply a Logback pattern string, which they used in the Logback config, to the Log Viewer via a URL parameter or manually without any conversion. 

On the other hand, we're implementing a generic Jsonl decoder / formatter, which doesn't necessarily need to bind with Logback. Let's repurpose this method to fix the date format strings whenever they don't make sense to Day.js. What do you think?

junhaoliao committed:
Apply suggestions from code review

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

junhaoliao committed:
Refactor variable name for clarity in StateContextProvider.

junhaoliao committed:
Apply suggestions from code review

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

junhaoliao committed:
Refactor JsonlDecoder and create new LogbackFormatter class.

junhaoliao committed:
Update decoder interface comment.

junhaoliao committed:
Add return type to #extractLogLevel method.

junhaoliao commented on new-log-viewer/src/typings/decoders.ts:
```diff
@@ -0,0 +1,53 @@
+ */
+type DecodeResultType = [string, number, number, number];
+
+interface Decoders {
```
definitely

junhaoliao committed:
Rename interface from "Decoders" to "Decoder".

junhaoliao committed:
Rename decode function parameter 'startIdx' to 'beginIdx'.

junhaoliao committed:
Simplify basename extraction in url parsing.

junhaoliao committed:
Validate file size; Add file size formatting utility.

junhaoliao committed:
Reformat code.

kirkrodrigues committed:
Refactor file.ts

kirkrodrigues committed:
Refactor http.ts

kirkrodrigues committed:
Refactor url.ts

kirkrodrigues committed:
Refactor worker.ts

kirkrodrigues committed:
Refactor js.ts

kirkrodrigues committed:
Refactor StateContextProvider.

kirkrodrigues committed:
Refactor MainWorker.ts

kirkrodrigues committed:
Rename FormatOptionsType -> FormatterOptionsType and DecodeOptionsType -> DecoderOptionsType.

kirkrodrigues committed:
Add missing semicolons.

kirkrodrigues committed:
Rename convertLogbackDateFormatToDayjs to convertDateTimeFormatterPatternToDayJs.

kirkrodrigues committed:
Add docstring for LogbackFormatterOptionsType.

kirkrodrigues committed:
Clean LogbackFormatter's unnecessary defaults.

kirkrodrigues committed:
Refactor decoders.ts

kirkrodrigues committed:
extractLogLevel -> parseLogLevel

kirkrodrigues committed:
Edit docstrings in JsonlDecoder.ts.

kirkrodrigues committed:
Refactor parseLogLevel to better handle non-string values.

kirkrodrigues committed:
extractDateFormat -> parseDateFormat, extractKeys -> parseKeys

kirkrodrigues committed:
Split extractAndFormatTimestamp into parseTimestamp and formatTimestamp to better separate concerns.

kirkrodrigues committed:
LogbackFormatter.ts: Use explicit casts to bypass typescript's unnecessary errors; Replace execs with matchAll.

kirkrodrigues committed:
input -> formatString, placeholder -> specifier, propName -> key, propValue -> value; Edit docstrings; Add TODO about variables which contain '%'.

kirkrodrigues committed:
getRangeFrom -> getCursorRange

kirkrodrigues committed:
startLogEventNum -> beginLogEventNum

kirkrodrigues committed:
Bug-fix: Off by one error.

kirkrodrigues committed:
LogFileManager.ts: Edit docstrings.

kirkrodrigues commented on new-log-viewer/src/utils/file.ts:
How about calling this units.ts?

kirkrodrigues reviewed (changes requested):
Nice work. As discussed offline:

* Let's pass an initial set of decoder options into the decoder constructor so that the decoder is always in a valid state (and we don't have to check as much when calling its methods).
  * We would also need to pass the initial decoder options when loading the file.
* Let's define the decoders' constructors agnostic to the interface of the future CLP FFI class (with the plan being to wrap that class in a Typescript class that can conform to our the current decoders' constructors).
* Let's count the number of log events in the decoder constructor, leaving `buildIdx` as an index-building method (if it's not already been built); this means `buildIdx` needs to take a range similar to `decode`.
  * The count should include invalid log events and in the future, we should display undecodable log events specially to the user.
* Let's make `loadFile` into `LogFileManager`'s constructor since we don't re-use the class' state between log files.

junhaoliao commented on new-log-viewer/src/utils/file.ts:
```diff
@@ -0,0 +1,43 @@
+    const units = useSiUnits ?
+        SI_UNITS :
+        IEC_UNITS;
+    /* eslint-disable no-magic-numbers */
```
```suggestion
```
Numbers 10 and 3 have been added into an exception list in `eslint-config-yscope` 0.0.30 or above: https://github.com/y-scope/eslint-config-yscope/commit/515b6dd15daae18a08c85e5e7a147159cf5b6216

Sorry for the confusion.

junhaoliao commented on new-log-viewer/src/utils/file.ts:
```diff
@@ -0,0 +1,43 @@
+    const divisor = useSiUnits ?
+        10 ** 3 :
+        2 ** 10;
+    /* eslint-enable no-magic-numbers */
```
```suggestion
```

junhaoliao commented on new-log-viewer/src/utils/file.ts:
```diff
@@ -0,0 +1,43 @@
+        2 ** 10;
+    /* eslint-enable no-magic-numbers */
+
+    // eslint-disable-next-line no-magic-numbers
```
```suggestion
```

junhaoliao commented on new-log-viewer/src/contexts/StateContextProvider.tsx:
```diff
@@ -79,14 +81,25 @@ const StateContextProvider = ({children}: StateContextProviderProps) => {
+                setBeginLineNumToLogEventNum(args.beginLineNumToLogEventNum);
+                const logEventNums = Array.from(args.beginLineNumToLogEventNum.values());
+
+                // Explicit cast since typescript thinks the last item of `logEventNums` can be
```
Hmm... Can `logEventNums` be an empty array due to programming errors / when there is no log event in the file?

kirkrodrigues replied:
True. In that case, should we set the log event num to 1 (rather than logging an error as we did previously)?

junhaoliao replied:
I agree

junhaoliao commented on new-log-viewer/src/typings/formatters.ts:
```diff
@@ -0,0 +1,45 @@
+
+type FormatterOptionsType = LogbackFormatterOptionsType;
+
+type TimestampAndMessageType = [number, string];
```
I was under the impression that returning an array is generally faster than returning an object. However, below benchmark code showed exactly the opposite:
```
// Function returning 2 items in an array
function returnArray() {
  return [1, 2];
}

// Function returning 2 items in an object
function returnObject() {
  return { a: 1, b: 2 };
}

let iterations = 100000;
let temp;

// Benchmark for array
console.time("Array Return");
for (let i = 0; i < iterations; i++) {
  temp = returnArray();
  const [first, second] = temp;
  console.log(first, second);
}
console.timeEnd("Array Return");

// Benchmark for object
console.time("Object Return");
for (let i = 0; i < iterations; i++) {
  temp = returnObject();
  const {a, b} = temp;
  console.log(a, b);
}
console.timeEnd("Object Return");
```
Output:
```
100000 1 2
Array Return: 801.222900390625 ms
100000 1 2
Object Return: 778.954833984375 ms
```

Let's use an object as the return type then.

junhaoliao commented:
> Let's count the number of log events in the decoder constructor, leaving buildIdx as an index-building method (if it's not already been built); this means buildIdx needs to take a range similar to decode.
The count should include invalid log events and in the future, we should display undecodable log events specially to the user.

Not sure if I get the right idea, but now buildIdx() sounds more appropriate being a helper of decode() when applicable. Let's further discuss (after I push all my local code) whether we want to keep the interface.

junhaoliao committed:
Update Decoder interface and restructure JsonlDecoder.

junhaoliao committed:
Refactor LogFileManager file load and creation workflow.

junhaoliao committed:
Add decoder options to JsonlDecoder constructor.

junhaoliao committed:
Improve error handling when there is 0 log event.

junhaoliao committed:
Use object instead of array for type TimestampAndMessageType.

junhaoliao committed:
Rename file.ts to units.ts and remove eslint-disables.

junhaoliao committed:
Upgrade eslint-config-yscope to version 0.0.31 wrap braceless statement with braces.

Commit:
Edit JsonlDecoder's docstrings.

kirkrodrigues committed:
Edit decoders.ts docstrings.

kirkrodrigues committed:
Reorder JsonlDecoder's methods to match decoders.ts

kirkrodrigues committed:
Minor edits to LogFileManager.ts

kirkrodrigues commented on new-log-viewer/src/services/decoders/JsonlDecoder.ts:
```diff
@@ -0,0 +1,145 @@
+
+        return {
+            numValidEvents: endIdx - beginIdx,
+            numInvalidEvents: 0,
```
We could technically count the number of invalid events in the constructor and then return something accurate here, right?

junhaoliao replied:
Implemented below:
1. If a line cannot be parsed as a JSON object, the error is logged, the line is skipped, and a placeholder is inserted into the results. In decode(), such lines are directly inserted without formatting.
2. Add tracking for invalid log events, which now contributes to the total count of invalid events returned by buildIdx().

kirkrodrigues commented on new-log-viewer/src/services/MainWorker.ts:
```diff
@@ -19,27 +35,46 @@ const postResp = <T extends WORKER_RESP_CODE>(
+
+                postResp(
+                    WORKER_RESP_CODE.NUM_EVENTS,
+                    {numEvents: LOG_FILE_MANAGER.numEvents}
```
Our guidelines in other languages are not to allow public access of members in non-trivial classes (since they could be set without calling the relevant methods). Any reason to make this public?

kirkrodrigues commented on new-log-viewer/src/services/LogFileManager.ts:
```diff
@@ -0,0 +1,200 @@
+        decoderOptions: DecoderOptionsType
+    ): Promise<LogFileManager> {
+        const mgr = new LogFileManager(pageSize);
+        await mgr.#loadFile(fileSrc);
```
As discussed offline, let's make this a static/non-class method, collect the results here, and then pass them into the `LogFileManager` constructor. That would also allow us to get rid of the "is #loadFile" called checks in other methods.

kirkrodrigues commented on new-log-viewer/src/services/LogFileManager.ts:
```diff
@@ -0,0 +1,200 @@
+     * @param decoderOptions Initial decoder options.
+     * @throws {Error} if #fileName or #fileData hasn't been init, or a decoder cannot be found.
+     */
+    #initDecoder = (decoderOptions: DecoderOptionsType): void => {
```
Fwiw, this method could be taken out of the class since it's really only meant to be called once during construction. That would also allow us to get of the `Unexpected usage` exception. That said, it's not a huge deal to leave it in the class.

junhaoliao replied:
Instead, the method is kept to retain the side effects, but now it returns the constructed decoder so that the `#decoder` instance can be definitely assigned in LogFileManager's constructor.

junhaoliao committed:
Fix condition check in JsonlDecoder.

junhaoliao committed:
Refactor numEvents as private field in LogFileManager.

junhaoliao committed:
Remove undefined logEvent check in JsonlDecoder.

junhaoliao committed:
Refactor LogFileManager constructor, make loadFile static and make #initDecoder to return the decoder so that it can be assigned in LogFileManager's constructor.

junhaoliao committed:
Enhance error handling within JsonlDecoder:
1. If a line cannot be parsed as a JSON object, the error is logged, the line is skipped, and a placeholder is inserted into the results. In decode(), such lines are directly inserted without formatting.
2. Add tracking for invalid log events, which now contributes to the total count of invalid events returned by buildIdx().

junhaoliao committed:
Remove outdated throws in docs.

junhaoliao committed:
Improve docs.

kirkrodrigues committed:
JsonlDecoder: Add TODO and comment.

kirkrodrigues committed:
LogFileManager: Fix docstring.

kirkrodrigues committed:
Pull general method loadFile out of LogFileManager.

kirkrodrigues reviewed (approved):
For the PR title, how about:

Add support for loading files, decoding them as JSONL, and formatting them using a Logback-like format string.

junhaoliao committed:
Replace invalid timestamp literal with constant in JsonlDecoder.
